### PR TITLE
Deployment 2

### PR DESCRIPTION
**Pull request by macat**

### CODE REVIEW

File: deployment/deployment.go [hunk 1]
```diff
@@ -53,22 +63,44 @@ type Deployment struct {
...
+		err := taskstep.step.Deploy()
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// collectSteps collects all the steps for the deployment from the playbook
```
Comment (bobbytables) [inline deployment/deployment.go]: As an fyi, for unexported functions you actually don't NEED to document them. I love when you do though!

Comment (cored) [inline deployment/deployment.go]: What does the steps prefix means?

File: deployment/deployment_test.go [hunk 2]
```diff
@@ -16,41 +16,79 @@ func init() {
...
-		Name: "Test deployment",
-		Meta: playbook.Meta{},
-		Vars: []string{"test"},
-		Tasks: []playbook.Task{
-			{
-				Name: "First step",
-				Manifests: []string{
-					"test",
+	cases := []struct {
+		Tasks    []playbook.Task
```
Comment (bobbytables) [inline deployment/deployment_test.go]: Can we add a name to our test tables and log them inside of the test table loop please? That way when things break we know exactly which test table entry failed.

File: deployment/deployment.go [hunk 3]
```diff
@@ -53,22 +73,54 @@ type Deployment struct {
...
+		if task.PodManifest != "" {
+			m := d.Manifests[task.PodManifest]
+			rendered := m.Execute(d.Variables)
+			object, err := deserialize(rendered)
+			if err != nil {
+				return steps, err
+			}
+			step := NewPodmanifestStep(object)
+			steps = append(steps, TaskStep{&task, step})
+		} else {
```
Comment (therealplato) [inline deployment/deployment.go]: caution, by my understanding we may see podManifest and Manifests in the same task, can anyone confirm/deny? if so, the else block will skip the Manifests

File: services/deployment.go [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+func (d *DeploymentService) Deploy(p playbook.Playbook, vars map[string]string) error {
+	return d.deployer.Deploy(p, vars)
+}
+
+// DefaultDeployer implements the Deployer interface
+type DefaultDeployer struct{}
+
+// Deploy finds the playbook's manifests and executes them with Kubernetes.
+func (dd *DefaultDeployer) Deploy(p playbook.Playbook, vars map[string]string) error {
+	MS := NewManifestService()
```
Comment (bobbytables) [inline services/deployment.go]: uppercase? What is this? :stuck_out_tongue:

File: deployment/deployment.go [hunk 5]
```diff
@@ -42,6 +55,13 @@ func init() {
 	if err != nil {
 		panic(err)
 	}
+
+	namespace = os.Getenv("KUBERNETES_NAMESPACE")
+}
+
+// Deployer declares something that can Deploy Deployments
+type Deployer interface {
+	Deploy(playbook.Playbook, map[string]string) error
```
Comment (bobbytables) [inline deployment/deployment.go]: What is `map[string]string` here? Vars?

File: deployment/kubernetes_test.go [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+			Name: "Step with 1 podmanifest file",
+			Tasks: []playbook.Task{
+				{
+					Name:        "First step",
+					PodManifest: "test",
+				},
+			},
+			Expected: 2,
+		},
+	}
```
Comment (therealplato) [inline deployment/kubernetes_test.go]: we should add test case for playbook with multiple tasks

File: services/deployment.go [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+}
+
+// Deploy deploys a playbook
+func (d *DeploymentService) Deploy(instance *broadway.Instance) error {
+
+	log.Println("---")
+	log.Printf("%+v", d.playbooks)
+	log.Println("---")
+	playbook := d.playbooks[instance.PlaybookID]
+
```
Comment (therealplato) [inline services/deployment.go]: add err if playbook was missing

File: services/deployment.go [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+	err := deployer.Deploy()
+	if err != nil {
+		log.Println(err)
+		instance.Status = broadway.StatusError
+	} else {
+		instance.Status = broadway.StatusDeployed
+	}
+	d.repo.Save(instance)
+
+	return nil
```
Comment (therealplato) [inline services/deployment.go]: maybe `return err` instead?

File: services/deployment.go [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+		return errors.New("Instance is being deleted already.")
+	}
+
+	instance.Status = broadway.StatusDeploying
+	d.repo.Save(instance)
+
+	err := deployer.Deploy()
+	if err != nil {
+		log.Println(err)
+		instance.Status = broadway.StatusError
```
Comment (macat) [inline services/deployment.go]: return error here.

File: services/deployment_test.go [hunk 10]
```diff
@@ -0,0 +1 @@
+package services
```
Comment (macat) [inline services/deployment_test.go]: add tests

File: services/manifest.go [hunk 11]
```diff
@@ -0,0 +1,98 @@
...
+// implementations
+type ManifestService struct {
+	rootFolder string
+	extension  string
+}
+
+// NewManifestService instantiates a ManifestService with a default rootFolder
+func NewManifestService() *ManifestService {
+	return &ManifestService{
+		rootFolder: "./manifests",
```
Comment (therealplato) [inline services/manifest.go]: use env var over default location

**Comment by therealplato** [comment]
+1

**Pull request merged**